doc/manual/performance-tips.rst

```diff
@@ -508,81 +964,48 @@ had we so desired.
 Taken to its extreme, pre-allocation can make your code uglier, so
 performance measurements and some judgment may be required.
 
+Inline functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Functions called within a loop are not always inlined.
 
-Avoid string interpolation for I/O
-----------------------------------
-
-When writing data to a file (or other I/O device), forming extra
-intermediate strings is a source of overhead. Instead of::
-
-    println(file, "$a $b")
-
-use::
-
-    println(file, a, " ", b)
-
-The first version of the code forms a string, then writes it
-to the file, while the second version writes values directly
-to the file. Also notice that in some cases string interpolation can
-be harder to read. Consider::
-
-    println(file, "$(f(a))$(f(b))")
-
-versus::
-
-    println(file, f(a), f(b))
-
-
-Optimize network I/O during parallel execution
-----------------------------------------------
-
-When executing a remote function in parallel::
-
-    responses = cell(nworkers())
-    @sync begin
-        for (idx, pid) in enumerate(workers())
-            @async responses[idx] = remotecall_fetch(pid, foo, args...)
-        end
-    end
-
-is faster than::
+    julia> f(x::Float64) = (x+1.2)^2
+    julia> function testn(f)
+               s = 0.0
+               for i = 1:10^7
+                   s += f(i/10^7)
+               end
+               s
+           end
 
-    refs = cell(nworkers())
-    for (idx, pid) in enumerate(workers())
-        refs[idx] = @spawnat pid foo(args...)
-    end
-    responses = [fetch(r) for r in refs]
+    julia> function test_inlined()
+               s = 0.0
+               for i = 1:10^7
+                   s += (i/10^7+1.2)^2
+               end
+               s
+           end
 
-The former results in a single network round-trip to every worker, while the
-latter results in two network calls - first by the ``@spawnat`` and the
-second due to the ``fetch`` (or even a ``wait``). The ``fetch``/``wait``
-is also being executed serially resulting in an overall poorer performance.
+    julia> @time testn(f)
+    elapsed time: 1.168110985 seconds (480138732 bytes allocated, 26.17% gc time)
```

Comment:
FastAnonymous allows one to pass functions-as-arguments and get them inlined. 
It doesn't change inlining behavior. It also doesn't force boxing the output 
of the function, which may be more important than the inlining.

You can prevent a function from being inlined using `@noinline`. Believe it or 
not, that sometimes gives you a performance advantage. So you could construct 
an example simply by adding `@noinline` in front of your function declaration.

On Tuesday, July 28, 2015 05:32:17 AM Matthieu Gomez wrote:
> > -The former results in a single network round-trip to every worker, while
> > the -latter results in two network calls - first by the ``@spawnat`` and
> > the -second due to the ``fetch`` (or even a ``wait``). The
> > ``fetch``/``wait`` -is also being executed serially resulting in an
> > overall poorer performance. +    julia> @time testn(f)
> > +    elapsed time: 1.168110985 seconds (480138732 bytes allocated, 26.17%
> > gc time)
> Ah, I did not get difference. Which aspect does FastAnonymous aim to solve?
> If I do the same correction in your `FastAnonymous` Readme example, I
> obtain the same timing for `testn` and `test_inlined` (albeit both time at
> 0.4 rather than 0.07 as reported in your test).
> 
> I committed a new version - it would be great the add a simple example for a
> non-inlined function if you have one in mind.
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/JuliaLang/julia/pull/12244/files#r35641110